mlir/lib/Transforms/SROA.cpp

```diff
@@ -0,0 +1,194 @@
+            dyn_cast<DestructibleAccessorOpInterface>(use.getOwner())) {
+      if (accessor.canRewire(slot, info.usedIndices, usedSafelyWorklist)) {
+        info.accessors.push_back(accessor);
+        continue;
+      }
+    }
+
+    // If it cannot be shown that the operation uses the slot safely, maybe it
+    // can be promoted out of using the slot?
+    scheduleAsBlockingUse(use);
+  }
+
+  SmallPtrSet<OpOperand *, 16> dealtWith;
+  while (!usedSafelyWorklist.empty()) {
+    MemorySlot mustBeUsedSafely = usedSafelyWorklist.pop_back_val();
+    for (OpOperand &subslotUse : mustBeUsedSafely.ptr.getUses()) {
+      if (dealtWith.contains(&subslotUse))
+        continue;
+      dealtWith.insert(&subslotUse);
+      Operation *subslotUser = subslotUse.getOwner();
+
+      if (auto memOp = dyn_cast<TypeSafeOpInterface>(subslotUser))
+        if (succeeded(memOp.ensureOnlyTypeSafeAccesses(mustBeUsedSafely,
+                                                       usedSafelyWorklist)))
+          continue;
+
+      // If it cannot be shown that the operation uses the slot safely, maybe it
+      // can be promoted out of using the slot?
+      scheduleAsBlockingUse(subslotUse);
+    }
+  }
+
+  SetVector<Operation *> forwardSlice;
+  mlir::getForwardSlice(slot.ptr, &forwardSlice);
+  for (Operation *user : forwardSlice) {
+    // If the next operation has no blocking uses, everything is fine.
+    if (!info.userToBlockingUses.contains(user))
+      continue;
+
+    SmallPtrSet<OpOperand *, 4> &blockingUses = info.userToBlockingUses[user];
+    auto promotable = dyn_cast<PromotableOpInterface>(user);
+
+    // An operation that has blocking uses must be promoted. If it is not
+    // promotable, destruction must fail.
+    if (!promotable)
+      return {};
+
+    SmallVector<OpOperand *> newBlockingUses;
+    // If the operation decides it cannot deal with removing the blocking uses,
+    // destruction must fail.
+    if (!promotable.canUsesBeRemoved(blockingUses, newBlockingUses))
+      return {};
+
+    // Then, register any new blocking uses for coming operations.
+    for (OpOperand *blockingUse : newBlockingUses) {
+      assert(llvm::is_contained(user->getResults(), blockingUse->get()));
+
+      SmallPtrSetImpl<OpOperand *> &newUserBlockingUseSet =
+          info.userToBlockingUses.getOrInsertDefault(blockingUse->getOwner());
+      newUserBlockingUseSet.insert(blockingUse);
+    }
+  }
+
+  return info;
+}
+
+void mlir::destructSlot(DestructibleMemorySlot &slot,
+                        DestructibleAllocationOpInterface allocator,
+                        OpBuilder &builder, MemorySlotDestructionInfo &info) {
+  OpBuilder::InsertionGuard guard(builder);
+
+  builder.setInsertionPointToStart(slot.ptr.getParentBlock());
+  DenseMap<Attribute, MemorySlot> subslots =
+      allocator.destruct(slot, info.usedIndices, builder);
+
+  llvm::SetVector<Operation *> usersToRewire;
+  for (auto &[user, _] : info.userToBlockingUses)
```

Comment:
```suggestion
  for (auto &[user, _] : info.userToBlockingUses)
```
I don't think `_` means an unnamed variable in C++, you're binding it to a variable called `_`. I would either admit that you're naming it, or fall back on `it.second`

Reply:
there was this make first range thing that Jakub suggested?